lib/helpers.py

```diff
@@ -10,437 +10,598 @@
+        """
+        mark_df = df[df.event_type == 'mark']
+
+        # we dont need the event_type anymore (to save memory)
+        mark_df = mark_df.drop(columns=['event_type'])
+
+        sorted_mark_df = mark_df.sort_values('ts')
+
+        deduplicated_mark_df = sorted_mark_df.drop_duplicates(['user_id'])
+
+        return deduplicated_mark_df
+
+    @staticmethod
+    def _calculate_ad_spend(df):
+        ad_spend_micros = df[(df.event_type == 'buying_conversion') & (df.ab_test_group == TEST)]['cost_eur'].sum()
+        return ad_spend_micros / 10 ** 6
+
+    @staticmethod
+    def _merge(marks_df, attributions_df):
+        """
+        `merge` joins the marked users with the revenue events and excludes any revenue event that happened before the
+        user was marked.
+        """
+        merged_df = pd.merge(attributions_df, marks_df, on='user_id')
+
+        return merged_df[merged_df.ts_x > merged_df.ts_y]
+
+
+class _CSVHelpers(object):
+    def __init__(self, customer, revenue_event, chunk_size=10 ** 6):
+        """
+        Internal class, containing technical read-write related methods and helpers
+        :param customer: Name of the customer the report is created for
+        :param revenue_event: An event which is going to be taken as a revenue event, e.g. "purchase"
+        :param chunk_size: How many lines should be taken for a single chunk during reads
+
+        :type customer: str
+        :type revenue_event: str
+        :type chunk_size: int
+        """
+        self.customer = customer
+
+        self.revenue_event = revenue_event
+
+        self.chunk_size = chunk_size
+
+        # columns to load from CSV
+        self.columns = dict()
+        self.columns[CSV_SOURCE_MARKS_AND_SPEND] = ['ts', 'user_id', 'ab_test_group', 'campaign_id', 'cost_eur',
+                                                    'event_type']
+        self.columns[CSV_SOURCE_ATTRIBUTIONS] = ['ts', 'user_id', 'partner_event', 'revenue_eur']
+
+    def read_csv(self, audience, source, date, chunk_filter_fn=None):
+        """
+        Helper to download CSV files, convert to DF and print time needed.
+        Caches files locally and on S3 to be reused.
+        """
+        now = datetime.now()
+
+        date_str = date.strftime('%Y%m%d')
+
+        filename = '{0}/{1}/{2}.csv.gz'.format(self._audience_data_path(audience), source, date_str)
+
+        # local cache
+        cache_dir = '{0}/{1}/{2}'.format(cache_folder, audience, source)
+        if not os.path.exists(cache_dir):
+            os.makedirs(cache_dir)
+
+        cache_file_name = '{0}/{1}.parquet'.format(cache_dir, date_str)
 
-        # remove events without a user id
-        filtered_chunk = filtered_chunk[filtered_chunk['user_id'].str.len() == 36]
+        # s3 cache (useful if we don't have enough space on the Colab instance)
+        s3_cache_file_name = '{0}/{1}/{2}/{3}.parquet'.format(self._audience_data_path(audience), source, cache_folder,
```

Comment:
Too long, maybe put .format() at a new line?

Reply:
Strictly speaking, nop: fits the 120 symbols length.

Hint: you can find the linter config in `setup.cfg` in `lib` folder :wink:

Reply:
I will still reformat it for readability though